validator_client/src/beacon_node_fallback.rs

```diff
@@ -0,0 +1,424 @@
+/// behaviour, where the failure of one candidate results in the next candidate receiving an
+/// identical query.
+pub struct BeaconNodeFallback<T, E> {
+    candidates: Vec<CandidateBeaconNode<E>>,
+    slot_clock: Option<T>,
+    spec: ChainSpec,
+    log: Logger,
+}
+
+impl<T: SlotClock, E: EthSpec> BeaconNodeFallback<T, E> {
+    pub fn new(candidates: Vec<CandidateBeaconNode<E>>, spec: ChainSpec, log: Logger) -> Self {
+        Self {
+            candidates,
+            slot_clock: None,
+            spec,
+            log,
+        }
+    }
+
+    /// Used to update the slot clock post-instantiation.
+    ///
+    /// This is the result of a chicken-and-egg issue where `Self` needs a slot clock for some
+    /// operations, but `Self` is required to obtain the slot clock since we need the genesis time
+    /// from a beacon node.
+    pub fn set_slot_clock(&mut self, slot_clock: T) {
+        self.slot_clock = Some(slot_clock);
+    }
+
+    /// The count of candidates, regardless of their state.
+    pub async fn num_total(&self) -> usize {
+        self.candidates.len()
+    }
+
+    /// The count of synced and ready candidates.
+    pub async fn num_synced(&self) -> usize {
+        let mut n = 0;
+        for candidate in &self.candidates {
+            if candidate.status(RequireSynced::Yes).await.is_ok() {
+                n += 1
+            }
+        }
+        n
+    }
+
+    /// The count of candidates that are online and compatible, but not necessarily synced.
+    pub async fn num_available(&self) -> usize {
+        let mut n = 0;
+        for candidate in &self.candidates {
+            if candidate.status(RequireSynced::No).await.is_ok() {
+                n += 1
+            }
+        }
+        n
+    }
+
+    /// Loop through any `self.candidates` that we don't think are online, compatible or synced and
+    /// poll them to see if their status has changed.
+    ///
+    /// We do not poll nodes that are synced to avoid sending additional requests when everything is
+    /// going smoothly.
+    pub async fn update_unready_candidates(&self) {
+        for candidate in &self.candidates {
+            // There is a potential race condition between having the read lock and the write
+            // lock. The worst case of this race is running `try_become_ready` twice, which is
+            // acceptable.
+            //
+            // Note: `RequireSynced` is always set to false here. This forces us to recheck the sync
+            // status of nodes that were previously not-synced.
+            if candidate.status(RequireSynced::Yes).await.is_err() {
+                // There exists a race-condition that could result in `refresh_status` being called
+                // when the status does not require refreshing anymore. This deemed is an
+                // acceptable inefficiency.
+                let _ = candidate
+                    .refresh_status(self.slot_clock.as_ref(), &self.spec, &self.log)
```

Comment:
Is it worth doing that in parallel? We are already in the async context so would only need to call select_all for the refresh_status futures.

Reply:
Good call, I've cherry-picked 02c228931342d239d03ca2e5349ae00edd8dca6f